### PR TITLE
Add README describing the software structure

### PR DESCRIPTION
**Pull request by ademuri**

### CODE REVIEW

File: software/generic/NetworkManager.hpp [hunk 1]
```diff
@@ -0,0 +1,40 @@
+#ifndef __NETWORK_MANAGER_H__
+#define __NETWORK_MANAGER_H__
+
+#include "Radio.hpp"
+#include "Types.hpp"
```
Comment (achew22) [inline software/generic/NetworkManager.hpp]: Should includes be relative to the project root?

File: software/generic/Radio.hpp [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+  uint8_t dataLength;
+
+  /**
+   * The raw application-layer data. The length is the max packet length (61)
+   * minus the size of the packet ID and type (3 bytes).
+   */
+  uint8_t data[PACKET_DATA_LENGTH];
+};
+
+inline bool operator==(const RadioPacket& lhs, const RadioPacket& rhs) {
```
Comment (achew22) [inline software/generic/Radio.hpp]: ooooohhhhh

File: software/generic/Radio.hpp [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+#include "Types.hpp"
+
+enum PacketType {
+  PING,
+};
+
+static const uint8_t PACKET_DATA_LENGTH = 58;
+
+struct RadioPacket {
```
Comment (achew22) [inline software/generic/Radio.hpp]: How does a struct differ from a class?

File: software/generic/Types.hpp [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+#include <cstring>
+
+// Provide the arduino random function
+uint16_t random(uint16_t max);
+
+uint16_t random(uint16_t min, uint16_t max);
+
+#else
+// Arduino
+#include "String.h"
```
Comment (achew22) [inline software/generic/Types.hpp]: Do you need to include "Random.h" or something from Arduino?

File: software/generic/test/NetworkManagerTest.cpp [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+  EXPECT_EQ(radio.getSentPacket(), nullptr);
+
+  // New id means it should rebroadcast
+  receivedPacket.packetId = 6789;
+  radio.setReceivePacket(&receivedPacket);
+  EXPECT_EQ(networkManager->receive(packet), true);
+  EXPECT_EQ(*radio.getSentPacket(), receivedPacket);
+
+  // Make sure it doesn't crash when exceeding the cache size
+  for (uint16_t i = 1; i < NetworkManager::kRecentIdsCacheSize * 2; i++) {
```
Comment (achew22) [inline software/generic/test/NetworkManagerTest.cpp]: Why not from i=0?

File: software/generic/NetworkManager.hpp (hunk 1 above)
Comment (ademuri) [reply software/generic/NetworkManager.hpp]: I don't think so, CMake doesn't expect that

File: software/generic/Types.hpp (hunk 4 above)
Comment (ademuri) [reply software/generic/Types.hpp]: Changed this to include Arduino.h, which contains most of the Arduino functions, etc.

File: software/generic/test/NetworkManagerTest.cpp (hunk 5 above)
Comment (ademuri) [reply software/generic/test/NetworkManagerTest.cpp]: Added a comment - an ID of 0 isn't valid

File: software/generic/Radio.hpp (hunk 3 above)
Comment (ademuri) [reply software/generic/Radio.hpp]: https://google.github.io/styleguide/cppguide.html#Structs_vs._Classes

tl;dr "Use a struct only for passive objects that carry data; everything else is a class"
